chia/data_layer/data_store.py

```diff
@@ -940,187 +1054,79 @@ async def get_proof_of_inclusion_by_key(
+            generation = row["generation"]
+            return int(generation)
+
+    async def write_tree_to_file(
         self,
+        root: Root,
+        node_hash: bytes32,
         tree_id: bytes32,
-        generation_begin: int,
-        max_generation: int,
-        num_generations: int = 10000000,
+        deltas_only: bool,
+        filename: str,
         *,
         lock: bool = True,
-    ) -> List[Union[InsertionData, DeletionData]]:
-        result: List[Union[InsertionData, DeletionData]] = []
-        async with self.db_wrapper.locked_transaction(lock=lock):
-            query_generation_begin = max(generation_begin - 1, 0)
-            query_generation_end = min(
-                generation_begin + num_generations, await self.get_tree_generation(tree_id=tree_id, lock=False) + 1
-            )
-            query_generation_end = min(query_generation_end, max_generation + 1)
-            roots = await self.get_roots_between(tree_id, query_generation_begin, query_generation_end, lock=False)
-            previous_root = None
-            for current_root in roots:
-                if previous_root is None:
-                    previous_root = current_root
-                    continue
-                else:
-                    operation = await self.get_single_operation(
-                        previous_root.node_hash, current_root.node_hash, current_root.status, lock=False
-                    )
-                    result.append(operation)
-                previous_root = current_root
+    ) -> None:
+        if node_hash == bytes32([0] * 32):
+            open(filename, "ab").close()
+            return
+
+        if deltas_only:
+            generation = await self.get_first_generation(node_hash, tree_id, lock=lock)
+            # Root's generation is not the first time we see this hash, so it's not a new delta.
+            if root.generation != generation:
+                return
+        node = await self.get_node(node_hash, lock=lock)
+        to_write = b""
+        if isinstance(node, InternalNode):
+            await self.write_tree_to_file(root, node.left_hash, tree_id, deltas_only, filename, lock=lock)
+            await self.write_tree_to_file(root, node.right_hash, tree_id, deltas_only, filename, lock=lock)
+            to_write = bytes(SerializedNode(False, bytes(node.left_hash), bytes(node.right_hash)))
+        else:
+            assert isinstance(node, TerminalNode)
+            to_write = bytes(SerializedNode(True, node.key, node.value))
 
-        return result
+        with open(filename, "ab") as writer:
+            writer.write(len(to_write).to_bytes(4, byteorder="big"))
+            writer.write(to_write)
 
     async def subscribe(self, subscription: Subscription, *, lock: bool = True) -> None:
         async with self.db_wrapper.locked_transaction(lock=lock):
+            servers_ip = json.dumps(subscription.ip)
+            servers_port = json.dumps(subscription.port)
```

Comment:
I felt it's more natural given how we update the subscriptions - in the DB case we'd have to delete everything each time and then reinsert the new data.

Reply:
We can't just add a subscription or remove one through the RPC?  We have to read them all out, modify as we want, then write them all back?

Reply:
we'll probably need priorities in this case - i.e. first the main server(s), then the mirrors, then the mirrors of the mirrors etc. I feel how to implement this depends a lot of what CW will do - i.e. if it'll manage all of that and just paste our list, or if it'll take them one by one, so it's probably worth discussing the design further in this case.

Reply:
Sounds good.  I don't know what is best but I'll mention the idea of associating priority numbers to each subscription instead of just being order based.  Anyways, @emlowe I guess this needs to get added for discussion and tracking.